### PR TITLE
Implement Window Controls Overlay

### PR DESCRIPTION
**Pull request by FeroTheFox**
Fixes: #7199 #8650 #6278 #4292 (on Windows)

Hey @Eugeny, I've daily-driven Tabby for a little while and I love its customizability. It reminds me of Vivaldi in a lot of ways! This is my first PR here and I'd love to try to fix a long-standing issue with the tab bar location and the window controls.

This PR aims to mitigate this issue by replacing the custom window controls with the Window Controls Overlay API that Electron provides.

Specifically, my code:
- Updates Electron to the latest stable (from version 22 to version 25) to fix an issue where overlay buttons with transparennt backgrounds have issues with hover styles (see https://github.com/electron/electron/pull/38693). This does not seem to affect any other functionality in Tabby - all packages appear to work fine and no obvious breaking changes were observed. I've been able to build and run the app perfectly fine.
- Adds new styles to compensate for the space taken up by the controls. Extra padding and adjustments are needed to prevent clipping.

## Known issues
I've tested this throroughly on Windows 11, but there's some other issues with this implementation that need to be addressed in order for this to be considered production-ready. 

### Issue 1
- Currently, this is only feasibly usable dark themes. The background color of the overlayed buttons are transparent to not clash with any themes, but the icon colors are hardcoded to white.

here's a visual example:
image

*Dark themes work flawlessly, however in lighter themes...*

image
It is impossible to know the window buttons exist, even when hovering directly over them.

We'd need to implement a way to pass the currently applied theme's colors to Electron (probably via IPC) and update the colors of the window controls accordingly so that they don't cause issues like this.

### Issue 2
- the settings button is obscured when the window frame is set to thin and tabs location is on top.
image

### Issue 3
- I did not consider Linux heavily when I made this PR. Linux users will still see a broken window controls experience when not using the native titlebar. This should probably be set by default if a user is on Linux.

Other than that, I'm proud of how this turned out!
electron_un4bK8oc2L

### CODE REVIEW

**Comment by Eugeny** [comment]
Thank you! Might be a stupid question, but what are the advantages of WCO besides it being native? 

PS issue 1: you'd need to watch for config changes and respond to them in `window.ts` same way it's done for window vibrancy settings right now. Issue 2 is just a missing margin somewhere now that the window controls are not in the DOM tree.

**Comment by FeroTheFox** [comment]
Not a stupid question at all! That's exactly its advantage. It unifies window control implementations across macOS and Windows. IIRC setting `frame` to false on macOS defaults to WCO. On Windows 11, it lets people use snap layouts

**Comment by FeroTheFox** [comment]
@Eugeny Pinging for the latest update - the WCO colors now respect the theme that you're using :)

https://github.com/Eugeny/tabby/assets/52982404/7ea77977-5a90-41fc-bace-e8ce9e15aab8

**Comment by FeroTheFox** [comment]
After a day of using a built version of the last commit here, I can say it's working quite smoothly now. Not sure how to resolve the CI errors, but I'm sure it's nothing a re-run can't fix. No rush to merge or anything :)

File: app/lib/window.ts [hunk 1]
```diff
@@ -94,12 +94,19 @@ export class Window {
         } else {
             if (process.platform === 'darwin') {
                 bwOptions.titleBarStyle = 'hidden'
+                // If not macOS and native appearance is not toggled, use WCO.
+            } else {
```
Comment (Eugeny) [inline app/lib/window.ts:98]: Does WCO work on Linux? This code path implies it does, the comment in [titleBar.component.scss](https://github.com/Eugeny/tabby/pull/8707/files#diff-d148ef5900e1ff0d7deaf652d073de91b29f7bf9c814006c71fda13d72acc853) implies it doesn't.

File: app/lib/window.ts [hunk 2]
```diff
@@ -94,12 +94,19 @@ export class Window {
...
+                bwOptions.titleBarStyle = 'hidden'
```
Comment (Eugeny) [inline app/lib/window.ts:99]: can be moved out of the `if`

File: app/lib/window.ts [hunk 3]
```diff
@@ -94,12 +94,19 @@ export class Window {
...
+                bwOptions.titleBarOverlay = {
+                    color: '#00000000',
+                }
             }
         }
 
         if (process.platform === 'darwin') {
             this.window = new BrowserWindow(bwOptions) as GlasstronWindow
         } else {
+            // this.window = new BrowserWindow(bwOptions) as GlasstronWindow
```
Comment (Eugeny) [inline app/lib/window.ts:109]: Please clean up commented code across the PR

File: tabby-core/src/components/appRoot.component.pug [hunk 4]
```diff
@@ -1,16 +1,20 @@
 title-bar(
   *ngIf='ready && !hostWindow.isFullscreen && config.store.appearance.frame == "full" && config.store.appearance.dock == "off"',
   (dblclick)='hostWindow.toggleMaximize()',
+  [class.hide-controls]='hostApp.platform !== Platform.Linux && !hostWindow.isFullscreen',
```
Comment (Eugeny) [inline tabby-core/src/components/appRoot.component.pug:4]: should be controlled via an @Input on TitleBarComponent, not a class

**Review by Eugeny — CHANGES_REQUESTED** [review]
Added a few comments, otherwise LGTM :v:

File: app/lib/window.ts (hunk 1 above)
Comment (FeroTheFox) [reply app/lib/window.ts:98]: [It doesn't](https://www.electronjs.org/docs/latest/tutorial/window-customization#window-controls-overlay-macos-windows), but good catch, will fix

File: tabby-core/src/components/appRoot.component.pug (hunk 4 above)
Comment (FeroTheFox) [reply tabby-core/src/components/appRoot.component.pug]: Done in the latest commit, window controls in the titlebar component is an ngIf now ✨

**Comment by FeroTheFox** [comment]
I believe that resolves your comments :)

**Comment by Eugeny** [comment]
Tested on all 3 platforms and it works nicely - well done! 

@all-contributors add @FeroTheFox for code

**Comment by Clem-Fern** [comment]
Hi @Eugeny, @FeroTheFox :)
I hope you two doing well.

I encountered an error while working on tabby with the master branch and I think it has something to do with this PR (I'm on Linux):
```
index.ts:173 Unhandled exception: TypeError: this.window.setTitleBarOverlay is not a function
    at IpcMainImpl.<anonymous> (/home/clem/git/tabby/app/dist/webpack:/tabby/lib/window.ts:397:25)
    at IpcMainImpl.emit (node:events:513:28)
    at IpcMainImpl.emit (node:domain:489:12)
    at WebContents.<anonymous> (node:electron/js2c/browser_init:2:88498)
    at WebContents.emit (node:events:513:28)
    at WebContents.emit (node:domain:489:12)
```

https://github.com/Eugeny/tabby/blob/0101ffddb65b6fac3553499d7e2d113f31d23ff1/app/lib/window.ts#L390-L403

After a quick look in the electron [doc](https://www.electronjs.org/fr/docs/latest/api/browser-window#winsettitlebaroverlayoptions-windows), it seems that the function `setTitleBarOverlay` is available on Windows platform only.

**Comment by Eugeny** [comment]
Hey, normally when Electron docs refer to functions as only available on platform X, that means it's a noop on other platforms (also supported by the electron TS typings). I've seen it too, but the error by itself is harmless as the setTitleBarOverlay call is at the very end of the event handler.

I'm waiting to see if it's going to be fixed in the electron updates and if it's not, I'll just replace it with `this.window['setTitleBarOverlay']?.(...`

**Comment by Clem-Fern** [comment]
Oh I see, I was not sure what was the impact exactly so, in doubt, i preferred report this.
Thank's for your insight anyway ;)

**Pull request merged**